README.md
# Table Master Parser

[![Travis build status](https://travis-ci.org/richardschneider/table-master-parser.svg)](https://travis-ci.org/richardschneider/table-master-parser)
[![Coverage Status](https://coveralls.io/repos/github/richardschneider/table-master-parser/badge.svg?branch=master)](https://coveralls.io/github/richardschneider/table-master-parser?branch=master)
[![npm version](https://badge.fury.io/js/table-master-parser.svg)](https://badge.fury.io/js/table-master-parser) 

The [Blue Chip Bridge Table Manager Protocol](http://www.bluechipbridge.co.uk/protocol.htm), allows bridge programs to communicate with each other in order to play bridge. ASCII messages (terminated with a CRLF) are exchanged over TCP/IP.  

This package parses the ASCII message and produces a javascript object representation commonly called an [Abstract Syntax Tree](https://en.wikipedia.org/wiki/Abstract_syntax_tree) (AST).

Use the [online parser](https://unpkg.com/table-master-parser/dist/doc/parser.html) to see the results of parsing a message.

[Documentation](https://unpkg.com/table-master-parser/dist/doc/index.html) on each message is automatically produced by 
this package.
The [change log](https://github.com/richardschneider/table-master-parser/releases) is automatically produced with
the help of [semantic-release](https://github.com/semantic-release/semantic-release).

## Getting started

**table-master-parser** is available for [Node.js](https://nodejs.org) and most modern browsers.  If you want to know if your currremt browser is compatible, run the [online test suite](https://unpkg.com/table-master-parser/test/index.html).

Install with [npm](http://blog.npmjs.org/post/85484771375/how-to-install-npm)

    > npm install table-master-parser --save

### Node

Include the package

    var tm = require('table-master-parser')
    
Parse a message

    tm.parse('South plays AS', (e, m) => console.log(m)))
    
Produces

    {   kind: 'play', 
        card: 'AS',
        seat: 'S'
    }

### Browser

Include the package from your project

    <script src="./node_modules/table-master-parser/dist/table-master-parser.min.js" type="text/javascript"></script>

or from the [unpkg CDN](https://unpkg.com)

    <script src="https://unpkg.com/table-master-parser/dist/table-master-parser.min.js"></script>

This will provide `tableMasterParser` as a global object, or `define` it if you are using [AMD](https://en.wikipedia.org/wiki/Asynchronous_module_definition).


CHANGELOG.md
The [change log](https://github.com/richardschneider/table-master-parser/releases) is automatically produced with
the help of [semantic-release](https://github.com/semantic-release/semantic-release).


test/parser.js
'use strict';

var tm = require('..');
require('should');

describe('Parser', () => {
    it('should parse a string', done => {
        tm.parse('Connecting "my great team" as South using protocol version 18', done);
    });

    it('should return an error on an unparsable message', done => {
        tm.parse('Connecting team as South using protocol version 18', (err, ast) => { //missing quotes around the team name.
            err.should.have.property('message');
            done();
        });
    });

    describe('AST', () => {
        it('should have the kind of message', done => {
            let msg = 'Connecting "my great team" as South using protocol version 18';
            tm.parse(msg, (err, ast) => {
                ast.should.have.property('kind');
                done();
            });
        });

        it('should have specific properties of the message', done => {
            let msg = 'Connecting "my great team" as South using protocol version 18';
            tm.parse(msg, (err, ast) => {
                ast.should.have.property('kind', 'connect');
                ast.should.have.property('teamName', 'my great team');
                ast.should.have.property('seat', 'S');
                ast.should.have.property('protocolVersion', 18);
                done();
            });
        });

        it('should have specific properties of the bid message', done => {
            let msg = 'east bids 3NT';
            tm.parse(msg, (err, ast) => {
                ast.should.have.property('kind', 'bid');
                ast.should.have.property('seat', 'E');
                ast.should.have.property('bid', '3NT');
                done();
            });
        });

        it('should have specific properties of the cards message', done => {
            let msg = "north's cards : S AKJ43. H -. D 6543. C QJT9.";
            tm.parse(msg, (err, ast) => {
                if(err) {
                    done(err);
                }
                ast.should.have.property('kind', 'cards');
                ast.should.have.property('seat', 'N');
                ast.should.have.property('cards').and.length(13);
                done();
            });
        });

        it('should have specific properties of the alterable bid message', done => {
            let msg = "north bids 1NT Alert. 12 to 14 total points";
            tm.parse(msg, (err, ast) => {
                if(err) {
                    done(err);
                }
                ast.should.have.property('kind', 'bid');
                ast.should.have.property('seat', 'N');
                ast.should.have.property('bid', '1NT');
                ast.should.have.property('alert', '12 to 14 total points');
                done();
            });
        });

    });

});


gulpfile.js
'use strict';

var gulp = require('gulp');
var rename = require('gulp-rename');
var jshint = require('gulp-jshint');
var mocha = require('gulp-spawn-mocha');
var browserify = require('browserify');
var uglify = require('gulp-uglify');
var coveralls = require('gulp-coveralls');
var source = require('vinyl-source-stream');
var buffer = require('vinyl-buffer');
var glob = require('glob');
var mochaPhantomJS = require('gulp-mocha-phantomjs');
var babel = require('babelify');

var DEBUG = process.env.NODE_ENV === 'debug',
    CI = process.env.CI === 'true';

var paths = {
    test: ['./test/**/*.js', '!test/{temp,temp/**}'],
    source: ['./lib/*.js', './bin/*'],
    doc: ['./doc/*.html', './doc/*.css']
};
paths.lint = paths.source.concat(paths.test);

var plumberConf = {};

if (CI) {
    plumberConf.errorHandler = function(err) {
        throw err;
  };
}

gulp.task('lint', function() {
  return gulp.src(paths.lint)
    .pipe(jshint())
    .pipe(jshint.reporter('jshint-stylish'))
    .pipe(jshint.reporter('fail'));
});

gulp.task('istanbul', function () {
  return gulp.src(paths.test, {read: false})
    .pipe(mocha({
      debugBrk: DEBUG,
      R: 'spec',
      istanbul: !DEBUG
    }));
});

gulp.task('test-browser', ['dist'], function () {
    return gulp.src('test/runner.html')
        .pipe(mochaPhantomJS({reporter: 'min'}));
});

gulp.task('docs', function() {
    require('./doc/make').run();    // dynamic documentation
    return gulp.src(paths.doc)      // static documentation
        .pipe(gulp.dest('./dist/doc'));
});

gulp.task('dist-lib', function() {
    return browserify('./index.js', { standalone: 'tableMasterParser'})
        .transform(babel, {presets: ['es2015']})
        .ignore('fs')
        .ignore('stream')
        .bundle()
        .pipe(source('table-master-parser.js'))
        .pipe(gulp.dest('./dist/'))
        .pipe(buffer())
        .pipe(uglify())
        .pipe(rename({suffix: '.min'}))
        .pipe(gulp.dest('./dist/'));
});

gulp.task('dist-test', function (cb) {
  glob('./test/**/*.js', {}, function (err, files) {
    var b = browserify({standalone: 'spec'});
    files.forEach(function (file) {
        b.add(file);
    });
    b
        .transform(babel, {presets: ['es2015']})
        .ignore('fs')
        .ignore('stream')
        .bundle()
        .pipe(source('table-master-parser.spec.js'))
        .pipe(gulp.dest('./dist'))
        .pipe(buffer())
        .pipe(uglify())
        .pipe(rename({suffix: '.min'}))
        .pipe(gulp.dest('./dist/'))
        .on('end', function() { cb(); });
  });
});

gulp.task('coverage', function () {
  return gulp.src('./coverage/lcov.info')
    .pipe(coveralls());
});

gulp.task('test',    ['lint', 'istanbul']);
gulp.task('dist',    ['docs', 'dist-lib', 'dist-test']);
gulp.task('ci',      ['test', 'test-browser', 'dist']);
gulp.task('default', ['test']);


lib/grammar.js
'use strict';

var ll = require("ll.js");

var g = new ll.Grammar();
var p = ll.Factory;

/* Helpers */
function push(value) {
    return (stack, cb) => {
        stack.push(value);
        cb();
    };
}

function setExample(name, text) {
    g.rules[name].example = (grammar, cb) => cb(null, text);
}

/* Create an AST node from the parser stack. */
function ast(kind) {
    let properties = Array.prototype.slice.call(arguments, 1).reverse();
    return (stack, cb) => {
        let node = { kind: kind };
        properties.forEach(name => {
            if (name.endsWith('?')) {
                name = name.slice(0, -1);
                let o = stack[stack.length - 1];
                if (typeof o === 'object' && name in o) {
                    node[name] = stack.pop()[name];
                }
            } else {
                node[name] = stack.pop();
            }
        });
        stack.push(node);
        cb();
    };
}

/* Tokens */
g.addRule('port', p.number());
g.addRule('protocolVersion', p.number());
g.addRule('teamName', p.string());

/* Expressions */
g.addRule('seat', p.either(
    p.keyword('north').op(push('N')),
    p.keyword('south').op(push('S')),
    p.keyword('east').op(push('E')),
    p.keyword('west').op(push('W'))
    )
);

g.addRule("seat'", p.either(
    p.keyword("north's").op(push('N')),
    p.keyword("south's").op(push('S')),
    p.keyword("east's").op(push('E')),
    p.keyword("west's").op(push('W'))
    )
);

g.addRule('suit', p.either(
    p.keyword('S').op(push('S')),
    p.keyword('H').op(push('H')),
    p.keyword('D').op(push('D')),
    p.keyword('C').op(push('C'))
    )
);

g.addRule('vulnerable', p.either(
    p.keyword('Neither vulnerable').op(push('Nil')),
    p.keyword('N/S vulnerable').op(push('NS')),
    p.keyword('E/W vulnerable').op(push('EW')),
    p.keyword('Both vulnerable').op(push('All'))
    )
);

g.addRule('card', p.regex(/^[AKQJT98765432][SHDC]/, 'KH'));

g.addRule('scards', p.either(
    p.keyword('-', true),
    p.regex(/^[AKQJT98765432]+/, 'AJ43')
));
g.addRule('hand',
    p.kleene(
        p.other('someCards')
    ).op((stack, cb) => {
        let cards = [];
        while (stack.length > 0) {
            let some = stack[stack.length - 1];
            if (Array.isArray(some)) {
                Array.prototype.push.apply(cards, stack.pop().reverse());
            } else {
                break;
            }
        }
        stack.push(cards.reverse());
        cb();
    })
);
setExample('hand', 'S AKJ43. H -. D 6543. C QJT9.');

g.addRule('someCards',
    p.sequence(
        p.other('suit'),
        p.other('scards'),
        p.keyword('.')
    ).op((stack, cb) => {
        var ranks = stack.pop();
        var suit = stack.pop();
        if (ranks === '-') {
            stack.push([]);
        } else {
            var cards = Array.from(ranks).map(r => r + suit);
            stack.push(cards);
        }
        cb();
    })
);

/* Messages */
let messages = [];
let otherMessages = [];
messages.add = function(name, term) {
    g.addRule(name, term);
    let rule = g.rules[name];
    this.push(rule);
    otherMessages.push(p.other(name));
};

/*
 * Connections messages
 */
messages.add('connect',
    p.sequence(
        p.keyword('connecting'),
        p.other('teamName'),
        p.keyword('as'),
        p.other('seat'),
        p.keyword('using'),
        p.keyword('protocol'),
        p.keyword('version'),
        p.other('protocolVersion')
    ).op(ast('connect', 'teamName', 'seat', 'protocolVersion'))
);

messages.add('readyForTeams',
    p.sequence(p.other('seat'), p.keyword('ready'), p.keyword('for'), p.keyword('teams'))
    .op(ast('readyForTeams', 'seat'))
);

messages.add('readyToStart',
    p.sequence(p.other('seat'), p.keyword('to'), p.keyword('start'))
    .op(ast('readyToStart', 'seat'))
);

messages.add('teams',
    p.sequence(
        p.keyword('teams'), p.keyword(':'),
        p.keyword('N/S'), p.keyword(':'), p.other('teamName'),
        p.keyword('E/W'), p.keyword(':'), p.other('teamName'))
    .op(ast('teams', 'NS', 'EW'))
);

messages.add('seated',
    p.sequence(p.other('seat'), p.other('teamName'), p.keyword('seat'))
    .op(ast('seated', 'seat', 'teamName'))
);

messages.add('endOfSession',
    p.sequence(
        p.keyword('end'),
        p.keyword('of'),
        p.keyword('session'))
    .op(ast('endOfSession'))
);

/*
 * Dealing messages
 */
messages.add('startOfBoard',
    p.sequence(p.keyword('start'), p.keyword('of'), p.keyword('board'))
    .op(ast('startOfBoard'))
);
messages.add('readyForDeal',
    p.sequence(p.other('seat'), p.keyword('ready'), p.keyword('for'), p.keyword('deal'))
    .op(ast('readyForDeal', 'seat'))
);
messages.add('readyForCards',
    p.sequence(p.other('seat'), p.keyword('ready'), p.keyword('for'), p.keyword('cards'))
    .op(ast('readyForCards', 'seat'))
);
messages.add('deal',
    p.sequence(
        p.keyword('board'), p.number(), p.keyword('.'),
        p.keyword('dealer'), p.other('seat'), p.keyword('.'),
        p.other('vulnerable'))
    .op(ast('deal', 'board', 'dealer', 'vulnerable'))
);
messages.add('cards',
    p.sequence(
        p.other("seat'"),
        p.keyword('cards'),
        p.keyword(':'),
        p.other('hand')
    )
    .op(ast('cards', 'seat', 'cards'))
);

/*
 * Bidding messages
 */
messages.add('illegalBid',
    p.sequence(p.keyword('illegal'), p.keyword('bid'))
    .op(ast('illegalBid'))
);

g.addRule('call', p.either(
    p.keyword('passes').op(push('pass')),
    p.keyword('doubles').op(push('X')),
    p.keyword('redoubles').op(push('XX')),
    p.sequence(
        p.keyword('bids'),
        p.regex(/^[1-7]([SHDC]|NT)/, '1NT')
    ))
);

messages.add('bid',
    p.sequence(
        p.other('seat'),
        p.other('call'),
        p.optionally(
            p.sequence(
                p.keyword('Alert.'),
                p.regex(/.+/, '12 to 14 total points')
            ).op((stack, cb) => {
                stack.push({ alert: stack.pop() });
                cb();
            })
          )
    )
    .op(ast('bid', 'seat', 'bid', 'alert?'))
);
setExample('bid', 'south bids 1NT Alert. 12 to 14 total points');

messages.add('readyForBid',
    p.sequence(
        p.other('seat'),
        p.keyword('ready'),
        p.keyword('for'),
        p.other("seat'"),
        p.keyword('bid'))
    .op(ast('readyForBid', 'seat', 'from'))
);

/*
 * Playing messages
 */
g.addRule('trickNumber', p.number());
g.addRule('mmss', p.regex(/^\d+\:\d+/, '3:30'));
g.addRule('hhmmss', p.regex(/^\d+\:\d+\:\d+/, '1:10:30'));
messages.add('play',
    p.sequence(
        p.other('seat'),
        p.keyword('plays'),
        p.other('card'))
    .op(ast('play', 'seat', 'card'))
);
messages.add('readyForCard',
    p.sequence(
        p.other('seat'),
        p.keyword('ready'), 
        p.keyword('for'), 
        p.other("seat'"),
        p.keyword('card'),
        p.keyword('to'),
        p.keyword('trick'),
        p.other('trickNumber'))
    .op(ast('readyForCard', 'seat', 'player', 'trick'))
);
messages.add('readyForDummy',
    p.sequence(
        p.other('seat'),
        p.keyword('ready'), 
        p.keyword('for'), 
        p.keyword('dummy'))
    .op(ast('readyForDummy', 'seat'))
);
messages.add('dummyCards',
    p.sequence(
        p.keyword("dummy's"), 
        p.keyword('cards'), 
        p.keyword(':'), 
        p.other('hand'))
    .op(ast('dummyCards', 'seat', 'cards'))
);
messages.add('endOfBoard',
    p.sequence(
        p.keyword('timing'),
        p.keyword('-'),
        p.keyword('N/S'),
        p.keyword(':'),
        p.keyword('this'),
        p.keyword('board'),
        p.other('mmss'),
        p.keyword(','),
        p.keyword('total'),
        p.other('hhmmss'),
        p.keyword(','),
        p.keyword('E/W'),
        p.keyword(':'),
        p.keyword('this'),
        p.keyword('board'),
        p.other('mmss'),
        p.keyword(','),
        p.keyword('total'),
        p.other('hhmmss')
    )
    .op(ast('endOfBoard', 'nsBoardTime', 'nsTotalTime', 'ewBoardTime', 'ewTotalTime'))
);

/*
 * The root rule is one of the messaged 
 */
g.addRule('message', p.either.apply(p, otherMessages), true);

//console.log('grammar', g);
//g.dumpHTML((err, html) => console.log(html));
//g.example((err, s) => console.log(s));

module.exports = g;
